### PR TITLE
For nested create operations, perform authorization checks after connection

### LINKED ISSUE #3901: Missing authorization rule for ensuring/enforcing a specific connection on node creation
Following recent fix https://github.com/neo4j/graphql/pull/3890

> As `@authorization(filter)` can't impact node creations, we need to use `@authorization(validate)`, and I just realized the same issue occurred.

**Describe the bug**
Unable to ensure/enforce that a specific rule is valid on a connected node while creating the parent node, or in the example below unable to ensure that the post is being created by the logged-in user.

**Type definitions**
```graphql
type User {
    id: ID!
}

type Post @authorization(validate: [
    {
        operations: [CREATE]
        where: { node: { author: { id: "$jwt.sub" } } }
    }
]) {
    title: String!
    content: String!
    author: User! @relationship(type: "AUTHORED", direction: IN)
}
```

**To Reproduce**

Step 1, create user.
```graphql
mutation {
    createUsers(input: [{ id: "michel" }]) { users { id } }
}
```

Step 2, create post linked to previously created user => fail.
```graphql
mutation {
    createPosts(input: [{
        title: "Test1",
        content: "Test1",
        author: { connect: { where: { node: { id: "michel" } } } }
    }]) {
        posts { title }
    }
}
```
I got a `Neo4jGraphQLForbiddenError: Forbidden` where I should get the freshly created node.

**Expected behavior**
The post should be created successfully as all conditions are respected.

**System (please complete the following information):**

* OS: ArchLinux
* Version: "@neo4j/graphql": "4.1.1"
* Node.js version: v18.16.1

### PR DESCRIPTION
**Pull request by darrellwarde**
# Description

Authorization checks for `CREATE` operations were being checked immediately after node creation. However, some rules will require relationships to be in place to fulfil them.

So for nested create operations, perform the authorization checks _after_ the relationship has been created.

## Complexity

Complexity: Medium

# Issue

Closes #3901

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the Neo4j GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-3949.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 41 | 45 | 1134 |
|  aggregations.NestedAggregation | 16553 | 16553 | 81 | 162 | 2174 |
|  aggregations.AggregationWithWhere | 11979 | 11979 | 51 | 88 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 22116987 | 22116987 | 3370 | 3097 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 22116987 | 22116987 | 3095 | 3127 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 9734 | 9734 | 51 | 59 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 9937 | 9937 | 64 | 59 | 2174 |
|  aggregations.DeeplyNestedCount | 13070331 | 13070329 | 4788 | 5308 | 2008534 |
|  batch-create.BatchCreate | 4200 | 4200 | 143 | 171 | 600 |
|  batch-create.BatchCreateSmall | 77 | 77 | 65 | 85 | 11 |
|  connect.createAndConnect | 14424 | 14424 | 175 | 218 | 3003 |
|  connections.Connection | 14082 | 14082 | 62 | 84 | 2174 |
|  connections.NestedConnection | 45466 | 45472 | 130 | 295 | 4516 |
|  connections.ConnectionWithSort | 3284 | 3284 | 73 | 210 | 1040 |
|  connections.ConnectionWithSortAndCypher | 16325 | 16325 | 105 | 222 | 2174 |
|  create.SimpleMutation | 7 | 7 | 68 | 96 | 1 |
|  cypher-directive.TopLevelMutationDirective | 1135 | 1135 | 36 | 34 | 1134 |
|  delete.SimpleDelete | 19401 | 19401 | 253 | 242 | 1040 |
|  delete.NestedDeleteInUpdate | 19198 | 19198 | 200 | 259 | 1311 |
|  2925.SingleRelationshipFilter | 6468 | 6468 | 53 | 92 | 1040 |
|  2925.NestedSingleRelationshipFilter | 22900 | 22900 | 121 | 123 | 2174 |
|  2925.SingleRelationshipRequiredFilter | 5201 | 5201 | 44 | 51 | 1040 |
|  2925.NestedSingleRelationshipRequiredFilter | 9361 | 9361 | 68 | 75 | 1040 |
|  query.SimpleQuery | 3121 | 3121 | 24 | 28 | 1040 |
|  query.SimpleQueryWithRelationship | 16162 | 16162 | 49 | 66 | 2174 |
|  query.QueryWhere | 9712 | 9695 | 40 | 51 | 2166 |
|  query.SimpleQueryWithNestedWhere | 9890 | 9873 | 61 | 71 | 2166 |
|  query.Nested | 10096041 | 10096041 | 11224 | 10730 | 2008534 |
|  query.NestedWithFilter | 10074401 | 10074401 | 11025 | 11113 | 2004000 |
|  query.OrFilterOnRelationships | 42017 | 42776 | 212 | 253 | 1632 |
|  query.OrFilterOnRelationshipsAndNested | 35907 | 35634 | 249 | 325 | 1632 |
|  query.QueryWithNestedIn | 13627 | 13419 | 64 | 91 | 1291 |
|  query.NestedConnectionWhere | 9834 | 9834 | 62 | 99 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9851 | 9939 | 91 | 140 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 18996 | 19128 | 149 | 247 | 1552 |
|  query.NestedWithRelationshipSingleFilters | 3881 | 3881 | 213 | 477 | 1134 |
|  query.Fulltext | 80 | 80 | 74 | 81 | 16 |
|  query.FulltextWithNestedQuery | 587 | 587 | 73 | 74 | 84 |
|  sorting.SortMultipleTypes | 3515 | 3515 | 108 | 115 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 14464 | 14486 | 107 | 127 | 2174 |
|  sorting.SortOnNestedFields | 14082 | 14082 | 64 | 72 | 2174 |
|  sorting.SortDeeplyNestedFields | 43198 | 43198 | 106 | 126 | 4516 |
|  sorting.SortWithTopLevelCypher | 3396 | 3396 | 82 | 94 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 80 | 77 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 67 | 82 | 35 |
|  unions.NestedUnion | 410637 | 410637 | 297 | 300 | 33033 |
|  unions.NestedUnionWithMissingFields | 384611 | 384611 | 279 | 315 | 33033 |
|  update.NestedUpdate | 16143 | 16143 | 109 | 122 | 2002 |

</details>
        
Old Schema Generation: 48.462s
Schema Generation: 49.769s
Old Subgraph Schema Generation: 58.341s
Subgraph Schema Generation: 59.520s

File: packages/graphql/src/translate/create-create-and-params.ts [hunk 1]
```diff
@@ -357,14 +376,14 @@ function createCreateAndParams({
...
-        creates.push(`WHERE ${authorizationPredicates.join(" AND ")}`);
-    }
+    // if (authorizationPredicates.length) {
+    //     creates.push(`WITH ${withVars.join(", ")}`);
+    //     if (authorizationSubqueries.length) {
+    //         creates.push(...authorizationSubqueries);
+    //         creates.push(`WITH *`);
+    //     }
+    //     creates.push(`WHERE ${authorizationPredicates.join(" AND ")}`);
+    // }
```
Comment (angrykoala) [inline packages/graphql/src/translate/create-create-and-params.ts:386]: :eyes:

File: packages/graphql/tests/tck/issues/3901.test.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+ * limitations under the License.
+ */
+
+import { gql } from "graphql-tag";
+import type { DocumentNode } from "graphql";
+import { Neo4jGraphQL } from "../../../src";
+import { formatCypher, translateQuery, formatParams } from "../utils/tck-test-utils";
+import { createBearerToken } from "../../utils/create-bearer-token";
+
+describe("https://github.com/neo4j/graphql/issues/3901", () => {
```
Comment (angrykoala) [inline packages/graphql/tests/tck/issues/3901.test.ts:26]: Could this be tested in an integration test as well?

**Review by angrykoala — DISMISSED** [review]

**Review by angrykoala — APPROVED** [review]

File: packages/graphql/tests/integration/issues/3901.int.test.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        await cleanNodes(session, [Serie, Season, User]);
+        await session.close();
+    });
+
+    afterAll(async () => {
+        await driver.close();
+    });
+
+    test("should not raise cardinality error when connecting on create", async () => {
+        const createUser = /* GraphQL */ `
```
[the remark this answers is not in the document]
Comment (darrellwarde) [reply packages/graphql/tests/integration/issues/3901.int.test.ts:127]: I will apply this every soon probably. There is obviously the "code" `gql` tag, and in a lot of places we use `#graphql` inside the code block. However, I discovered that Prettier doesn't support that, and this label is what Prettier recognises. Nothing is simple!

**Pull request merged**